### PR TITLE
Enhance authorization for document creation workflow

### PR DESCRIPTION
**Pull request by nedalhy**
**Problem Description:**
When a document is created this is currently done by invoking an endpoint on the service (with a generic JWT token). The service will then create the document and return the ID of the document. The client is now expected to provide a new signed JWT for this new document.

This problem occur, when having an authorizaiton service that manages documents access control. It needs a way to confirm/validate that the user who created a document is the same user who is claiming to be creating the document against the authorizaiton service prior creating the ACL entry. 

**Proposal**
Modify the service to return a special JWT token (Creation token) in addition to the document ID. This token doesn't provide the user with any kind of access, but it will have the user's payload and documentId. E.g.:
```
{
  "documentId": "02f0c7fb-2e15-4d7f-82ea-b825ee65d778",
  "tenantId": "tenant_example",
  "user": {     // Optional user's payload
    "userId": "some_user",
    "creationId": "f9b7cda7-f5b9-4e55-aa81-91a7fc57ef3c" // Provided by the authorization service
  },
  "scopes": [],
  "iat": 1646662277,
  "exp": 1646662577,
  "ver": "1.0",
  "jti": "98b54f53-150e-493d-a084-dd77800e0974"
}
```

For making the usage of this token cleaner in the code,  the proposal extends `ITokenProvider ` with optional callback `documentPostCreateCallback` to consume this token, this callback is triggered immediately after creating the document, and could be used to call authorization service with the creation token and document id. E.g. 
```
 documentPostCreateCallback: async (documentId: string, creationToken: string) => {
      await authorizationService.createContainer(someAccessToken, documentId, creationToken);
  };
```

### CODE REVIEW

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:94]: These driver changes would not be released at the same time as the service changes, so token should be optional here, and `token===undefined` should be handled accordingly

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:107]: You can use `logger2.sendErrorEvent()` to log this

File: server/routerlicious/packages/services-utils/src/auth.ts
Comment (znewton) [inline server/routerlicious/packages/services-utils/src/auth.ts:62]: Utilizing `generateToken` from this same file would be preferable to duplicating the signing logic IMO

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (nedalhy) [reply packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:94]: Does it mean it need to handle temporary two return types `string` and `{id: string, token: string}`, until the service is deployed?

Comment (znewton) [reply packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:95]: Correct, yes. I believe you could check this with `if (typeof response === "string")`

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:114]: For the sake of future simplicity here, could we have the generateToken==false case return `{ id: string}`? That way, we can eventually remove the `| string` handling in the driver, and it will simplify the potential for adding more properties to the response body here.

This change would then require token to be expected to be possibly undefined in the driver. A little more complicated now for less complicated eventually 😄

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:95]: See other comment, but expecting token to be possibly undefined here would be great. Looks like the logic below can be left the same.

**Comment by znewton** [comment]
Awesome, thanks for addressing all the feedback, @nedalhy!

**Comment by nedalhy** [comment]
Thank you @znewton for your review. I see that there is one script failing `Move Non-Scoped Package fluid-framework` , do you perhaps know why is that ? 
Also, what would be the next steps in order to approve and merge the changes ?

**Comment by znewton** [comment]
> Also, what would be the next steps in order to approve and merge the changes ?

I believe @sharmakhushboo and/or @pradeepvairamani want to review this PR from a security standpoint. But the next step for you would be marking it as "Ready for Review" (currently a Draft). Then, after one of the above mentioned folks gives it the OK, we should be good to approve.

Also, have you validated that all of this works by running everything locally?

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:101]: I think the default here should be `this.tokenProvider.documentPostCreateCallback !== undefined`. Service generating a token takes a small amount of time, but it also increases the payload size which adds a little more time. For anyone not utilizing this feature, that additional latency is unnecessary.

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:122]: By catching the error here, if the `documentPostCreateCallback` throws an error (which I'm guessing would be the behavior if the token is invalid for the current user), then continuing will not be blocked. You could/should probably just remove this try..catch and event log if the intent is to fail here and prevent user from continuing.

**Review by znewton — COMMENTED** [review]
A few more comments

**Comment by nedalhy** [comment]
> I believe @sharmakhushboo and/or @pradeepvairamani want to review this PR from a security standpoint. But the next step for you would be marking it as "Ready for Review" (currently a Draft). Then, after one of the above mentioned folks gives it the OK, we should be good to approve.
> 
> Also, have you validated that all of this works by running everything locally?

I validated that the change with token generation and token provider callback works locally using an external application.
However, I couldn't find a sample in the repo that can be easily configured to use Routerlicious for testing, it seems like all of them are configured to use Tinylicious, but maybe i'm missing something.

Regarding backward compatibility, I assume if the tests are passing its sufficient to validate that, right ?

**Comment by znewton** [comment]
> Regarding backward compatibility, I assume if the tests are passing its sufficient to validate that, right ?

Unfortunately our tests are not currently sufficient to validate much. There are some instructions in the R11s README for [Running R11s](https://github.com/microsoft/FluidFramework/tree/main/server#running-the-reference-server) and [building your local changes](https://github.com/microsoft/FluidFramework/tree/main/server#running-the-reference-server). To run sample apps in the repo against R11s instead of T9s, you can simply run `npm run start:docker`. To run any app against local r11s, you can use the following config values:

```JS
{
  tenantId: "fluid",
  tenantSecret: "create-new-tenants-if-going-to-production",
  ordererUrl: "http://localhost:3003",
  storageUrl: "http://localhost:3001",
}
```

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:115]: My apologies, I asked you to convert this to an object. Would you mind switching it back to just the ID string? It slipped my mind that we need to remain backward compatible to old driver versions that can't handle an object response here 🤦‍♂️

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (znewton) [reply packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:122]: I would expect the callback failure to be handled by the user's implementation (e.g. adding a try..catch within their own implementation, that parses the error and rethrows if it is blocking)

**Comment by znewton** [comment]
I would also expect the [command shown here](https://github.com/microsoft/FluidFramework/tree/main/server#with-webpack-dev-server--the-webpack-fluid-loader-yo-fluid-output) to work for example apps (FF/examples/apps/*)

```
npx webpack-dev-server --config webpack.config.js --package package.json --env.mode docker
```

**Comment by nedalhy** [comment]
Thanks @znewton for sharing the instructions, I validated the changes against R11s and T9s using `clicked` sample. Everything seems to work as expected, I will mark the PR as "ready for review".

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:107]: I'd highly recommend leaving enough breadcrumbs in the code for next developer to be able to assess if it's now safe to remove back-compat code, and what code needs to be removed / modified. We have plenty of examples where it's not the case and code stays littered with dead code and comments like that.

File: packages/drivers/routerlicious-driver/src/tokens.ts
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/tokens.ts:54]: Can you please add to description why it's an async task, and what driver does (or does not) while awaiting it? I.e. is the creation process stalled while waiting for it? Is it expected that this callback will throw exceptions if something goes wrong and how the rest of pipeline will behave? (I.e. is file deleted in such case?) What's the purpose of this interaction (if it's not fire-and-forget)?

File: lerna-package-lock.json
Comment (nedalhy) [reply lerna-package-lock.json:1553]: Ohh, I thought I reverted changes on this file before, probably I introduced them in later commits.

File: packages/drivers/routerlicious-driver/src/tokens.ts
Comment (znewton) [reply packages/drivers/routerlicious-driver/src/tokens.ts:57]: +1 to noting that this will slow down the document creation process

**Review by pradeepvairamani — APPROVED** [review]

File: server/routerlicious/packages/services-utils/src/auth.ts
Comment (nedalhy) [reply server/routerlicious/packages/services-utils/src/auth.ts:51]: From the code it looks like `lifetime` is in [seconds](https://github.com/microsoft/FluidFramework/blob/d8f825ec284baf904912a2cbb82c87164ed463e8/packages/drivers/local-driver/src/auth.ts#L29). Nevertheless, It should be `5 * 60`, but I made a mistake when I changed this file to reuse `generateToken`.  
The idea is not to have long lifetime for creation token compared to the normal tokens, because ideally it should be used only once and immediately after document creation.

**Comment by pradeepvairamani** [comment]
> @znewton do you think we can merge this PR ? or we should wait for #9685 to be merged first?

@nedalhy Please feel free to merge this when the comments are addressed. You do not have to wait for #9685 to be merged.

**Comment by nedalhy** [comment]
@pradeepvairamani I don't have permissions to merge it, usually I ask @DLehenbauer to help us with merging PRs.

File: server/routerlicious/packages/services-utils/src/auth.ts
Comment (znewton) [reply server/routerlicious/packages/services-utils/src/auth.ts]: Thanks for checking me on that. You're totally right that it is in seconds

**Pull request merged**
